src/_pytest/cacheprovider.py

```diff
@@ -103,6 +103,14 @@ def set(self, key, value):
             with f:
                 self.trace("cache-write %s: %r" % (key, value))
                 json.dump(value, f, indent=2, sort_keys=True)
+        self._ensure_readme()
```

Comment:
This probably should be indented with the `with f` statement above; this way we ensure we have written the key/value and will be able to write the readme file as well.